### ISSUE TITLE
Inconsistent results when differentiating through a `torch.CvxpyLayer` repeatedly

### ISSUE THREAD

**Issue by PTNobel** [cvxgrp/cvxpylayers#119]
I'm computing the divergence of a function which includes a `CvxpyLayer` object as one phase of its computation.

I use two different strategies to differentiate through the program.

In the `v1` I differentiate each of the outputs with respect to `x` via `torch.backward` and select the relevant gradient.
In `v2` I differentiate each of the outputs with respect to only one input `xi`.
 The closed form expression for the divergence in this problem shows that `v2` is correct since the divergence should be an integer.

Notably, the output is the same in the first two partial derivatives, and then starts to diverge, so I suspect the issue has to do with internal state being modified inappropriately.

I believe this code is mathematically identical and should both output the same result. They do so when I use ADMM instead of CvxPy.

```python
import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer
import numpy as np
import torch

n = 10
m = 5

np.random.seed(n)
X = torch.Tensor(np.random.randn(n, m))
y_data = np.random.randn(n)

def generate_problem():
    beta = cp.Variable(m)
    y = cp.Parameter(n)
    obj = cp.Minimize(cp.sum_squares(y - X @ beta) / 2 + cp.pnorm(beta, p=1))
    prob = cp.Problem(obj, [])
    layer = CvxpyLayer(prob, parameters=[y], variables=[beta])
    return layer


def divergence_v1(x):
    x = torch.Tensor(x)
    x.requires_grad_(True)
    layer = generate_problem()
    sol = X @ layer.forward(x)[0]

    divergence = torch.Tensor((0.0,))
    for (i, gi) in enumerate(sol):
        gi.backward(retain_graph=True)
        divergence += x.grad[i]
        print(f"g_{i}/x_{i} = {x.grad[i]}")

    return divergence

def divergence_v2(x):
    split_data_into_leaves = []
    for xi in x:
        xi = torch.Tensor(xi)
        xi.requires_grad_(True)
        split_data_into_leaves.append(xi)
    x_for_eval = torch.stack(split_data_into_leaves)

    layer = generate_problem()
    sol = X @ layer.forward(x_for_eval)[0]

    divergence = torch.Tensor((0.0,))
    i= 0
    for (xi, gi) in zip(split_data_into_leaves, sol):
        gi.backward(inputs=[xi], retain_graph=True)
        divergence += xi.grad
        print(f"g_{i}/x_{i} = {xi.grad}")
        i+=1
    return divergence

print(f"{divergence_v1(y_data)=}")
np.random.seed(10)
X = torch.Tensor(np.random.randn(n, m))
y_data = torch.Tensor(np.random.randn(n))


print(f"{divergence_v2(y_data)=}")
```

Which generate output:

```
g_0/x_0 = 0.1500476896762848
g_1/x_1 = 0.023665698245167732
g_2/x_2 = 0.42122676968574524
g_3/x_3 = -0.10932222008705139
g_4/x_4 = 0.2735772132873535
g_5/x_5 = 0.12126564979553223
g_6/x_6 = 0.6079447269439697
g_7/x_7 = -0.25033044815063477
g_8/x_8 = -0.08230450749397278
g_9/x_9 = 0.04482615366578102
divergence_v1(y_data)=tensor([1.2006])
g_0/x_0 = 0.1500476896762848
g_1/x_1 = 0.013603110797703266
g_2/x_2 = 0.2239968478679657
g_3/x_3 = 0.4133704602718353
g_4/x_4 = 0.5792962908744812
g_5/x_5 = 0.06469085067510605
g_6/x_6 = 0.30013394355773926
g_7/x_7 = 0.10120414942502975
g_8/x_8 = 0.1321064978837967
g_9/x_9 = 0.021551402285695076
divergence_v2(y_data)=tensor([2.0000])
```

**Comment by PTNobel** [comment]
Good news; the issue was a bug in how I was using torch.autodiff, not in CvxpyLayers.

Divergence version 1 should be

```python



def divergence_v1(x):
    x = torch.Tensor(x)
    x.requires_grad_(True)
    layer = generate_problem()
    sol = X @ layer.forward(x)[0]

    divergence = torch.Tensor((0.0,))
    for (i, gi) in enumerate(sol):
        gi.backward(retain_graph=True)
        divergence += x.grad[i]
        print(f"g_{i}/x_{i} = {x.grad[i]}")
        x.grad = None

    return divergence
```
